### ISSUE TITLE
Performance of the different attention variants in the repo

### ISSUE THREAD

**Issue by jsmith1915** [lucidrains/sinkhorn-transformer#18]
Really great work!! 

I have a few questions. 

-Does attention sort net work better or at par with the simple sort net.
-Does attention sort net work well for the sortcut variant also.
-How does the sortcut variant perform when compared to vanilla sparse sinkhorn attention.

It would be very helpful if you could share some plots/numbers from your experiments comparing the performance of the different variants such as attention sort net, routing based attention etc.

**Comment by lucidrains** [comment]
@jsmith1915 thanks! in my toy tasks in the examples, I have found that the attention variant performs on par with the simple sorting network, but not any better in terms of accuracy. 

However, importantly, the attention sort net allows me to solve a problem where the sorting net is unable to generalize outside of the sequence length it is trained on. (the original sorting net consists of a linear network mapping `embedding dim -> number of buckets`). I used the attention matrix itself as the sorting matrix R, and therefore solved that issue. This is why, if you were to fix the hyperparameters to use the sinkhorn variant in the non-causal case, you will find that I assert the sequence length to be held constant.

**Comment by jsmith1915** [comment]
The attention sortnet is a very interesting improvement! Does replacing the Sinkhorn iterations with the softmax work well even with the simple sort net?

**Comment by lucidrains** [comment]
Yea, I was actually surprised that it does! It doesn't guarantee a permutative sorting matrix, but I found that it didn't matter. The algorithmic sort task the author was tackling in the paper probably benefitted the most from Sinkhorn, but in my use cases, a softmax actually worked fine. Count me surprised!

**Comment by lucidrains** [comment]
I have also found occasionally instability when training with Sinkhorn, and switching to the softmax made it less frequent

**Comment by jsmith1915** [comment]
Interesting results! From your experiments, did you observe any reduction in the instability when sortcut was used?

**Comment by lucidrains** [comment]
@jsmith1915 Yes, I think so actually. One of my theories is that the instability comes from abrupt transitions in the chosen key/value buckets during sorting. My recent change of allowing to sort multiple buckets to one query bucket was to address this.

**Comment by jsmith1915** [comment]
Thanks for the details!
